File: web/src/store/ws.ts
import {defineStore} from 'pinia';
import {WsMsg, WsState} from './interface';
import {UserStore} from "./user";
import {CONST} from "./constant";
import utils from "../utils/utils";
import {GlobalStore} from "./index";

export const WsStore = defineStore({
    id: 'WsState',
    state: (): WsState => ({
        ws: null,
        msg: {
            action: 0,
            data: {},
            type: "",
            uid: 0,
            rid: 0,
        },
        uid: -1,
        rid: -1,
        timeout: null,
        random: "",
        openNum: 0,
        listener: {},
    }),
    actions: {
        connection() {
            const globalStore = GlobalStore()
            const userStore = UserStore()
            //
            if (this.uid === userStore.info.id || !userStore.info.token) {
                return
            }
            //
            clearTimeout(this.timeout);
            if (this.ws) {
                this.ws.close();
                this.ws = null;
            }
            //
            let url = window.location.origin + (import.meta.env.VITE_API_URL as string) + "/ws";
            url = url.replace("https://", "wss://");
            url = url.replace("http://", "ws://");
            url += `?token=${userStore.info.token}&language=${globalStore.language}`;
            //
            const wgLog = true;
            const random = utils.randomString(16);
            this.random = random;
            //
            this.ws = new WebSocket(url);
            this.ws.onopen = async (e) => {
                wgLog && console.log("[WS] Open", e, utils.formatDate())
                this.openNum++;
            };
            this.ws.onclose = async (e) => {
                wgLog && console.log("[WS] Close", e, utils.formatDate())
                this.ws = null;
                //
                clearTimeout(this.timeout);
                this.timeout = setTimeout(() => {
                    random === this.random && this.connection();
                }, 3000);
            };
            this.ws.onerror = async (e) => {
                wgLog && console.log("[WS] Error", e, utils.formatDate())
                this.ws = null;
                //
                clearTimeout(this.timeout);
                this.timeout = setTimeout(() => {
                    random === this.random && this.connection();
                }, 3000);
            };
            this.ws.onmessage = async (e) => {
                wgLog && console.log("[WS] Message", e);
                const wsMsg: WsMsg = utils.jsonParse(e.data);
                this.msg = wsMsg;
                //
                if (wsMsg.action === CONST.WsOnline) {
                    // 连接成功
                    if (wsMsg.data.own === 1) {
                        this.rid = wsMsg.data.rid
                    }
                }
                Object.values(this.listener).forEach(call => {
                    if (typeof call === "function") {
                        try {
                            call(wsMsg);
                        } catch (err) {
                            wgLog && console.log("[WS] Callerr", err);
                        }
                    }
                });
            }
        },

        listener(name, callback) {
            if (typeof callback === "function") {
                this.listener[name] = callback;
            } else {
                this.listener[name] && delete this.listener[name];
            }
        },

        close() {
            if (this.ws) {
                this.ws.close();
                this.ws = null;
            }
        }
    },
});

File: app/interfaces/ws.go
package interfaces

import "github.com/gorilla/websocket"

type WebsocketReq struct {
	Token    string `json:"token"`
	Language string `json:"language"`
	Type     string `json:"type"`
}

type WsClient struct {
	Conn *websocket.Conn `json:"conn"`

	Ip   string `json:"ip"`   // 客户端IP
	Type string `json:"type"` // 客户端类型（如：user）
	Uid  string `json:"uid"`  // 客户端用户ID（会员ID）
	Rid  int32  `json:"rid"`  // 客户端序号ID（WebSocket ID）
}

type WsMsg struct {
	Action int `json:"action"` // 消息类型：1、上线；2、下线；3、消息
	Data   any `json:"data"`   // 消息内容

	Type string `json:"type"` // 客户端类型（如：user）
	Uid  string `json:"uid"`  // 客户端用户ID（会员ID）
	Rid  int32  `json:"rid"`  // 客户端序号ID（WebSocket ID）

	Md5    string `json:"md5"`
	Output string `json:"output"`
	Err    string `json:"err"`
}
